contracts/liquid-collateral/src/contract.rs

```diff
@@ -0,0 +1,1089 @@
+     - potential excessive principal amount is sent to position_created_address
+     - exact amount needed to be replenished is sent to principal_funds_owner
+   - in case there were not enough principal amount for replenish:
+     - send whatever principal amount is fetched to principal_funds_owner
+     - decrement principal amount needed
+     - update counteparty amount available in the auction
+     - start the auction
+*/
+pub fn end_round(deps: DepsMut, env: Env, _info: MessageInfo) -> Result<Response, ContractError> {
+    // Load the current state
+    let mut state = STATE.load(deps.storage)?;
+
+    let current_time = env.block.time;
+
+    // Check that the round is ended by checking that the round_id is less than the current round
+    if current_time < state.round_end_time {
+        return Err(ContractError::Std(StdError::generic_err(
+            "Round has not ended yet",
+        )));
+    }
+
+    // Create withdraw message
+    let withdraw_position_msg = MsgWithdrawPosition {
+        position_id: state.position_id.unwrap(),
+        sender: env.contract.address.to_string(),
+        liquidity_amount: state.liquidity_shares.clone().unwrap(),
+    };
+
+    // Query the claimable spread rewards
+    let spread_rewards = ConcentratedliquidityQuerier::new(&deps.querier)
+        .claimable_spread_rewards(state.position_id.unwrap())
+        .map_err(|_| ContractError::ClaimableSpreadRewardsQueryFailed {})? // Handle query errors
+        .claimable_spread_rewards;
+
+    // Query the claimable incentives
+    let incentives: ClaimableIncentivesResponse = ConcentratedliquidityQuerier::new(&deps.querier)
+        .claimable_incentives(state.position_id.unwrap())
+        .map_err(|_| ContractError::ClaimableIncentivesQueryFailed {})?;
+
+    // Save into state
+    state.position_rewards = Some(
+        fetch_all_rewards(
+            spread_rewards,
+            incentives.claimable_incentives,
+            incentives.forfeited_incentives,
+        )
+        .unwrap_or_default(),
+    );
+
+    // Wrap in SubMsg to handle response
+    let submsg = SubMsg::reply_on_success(withdraw_position_msg, 3);
+
+    // Save the updated state back to storage
+    STATE.save(deps.storage, &state)?;
+
+    Ok(Response::new()
+        .add_submessage(submsg)
+        .add_attribute("action", "withdraw_position"))
+}
+
+/*
+## End round bid
+ - can only be executed if auction is in progress
+ - bidder sends desired principal amount to the contract and request some amount of counterparty token
+ - bidder must replenish at least 1% of principal
+ - bidder cannot request more counterparty than contract has available
+ - bid is being saved in bids and in sorted bids
+ - in case the new bid makes principal amount be potentially replenished, unneeded (worst) bid/s will be kicked out
+ - if one or more bids are kicked out from sorted bids - bidders will be refunded and correct status of the bid will be saved
+*/
+pub fn end_round_bid(
```

Comment:
This name isn't great imo, because it's a noun. all other execute messages are verbs. I would call it
place_bid or add_bid or similar